OmsAgent/omsagent.py

```diff
@@ -46,100 +49,279 @@
+    return exit_code
 
 
 def uninstall(hutil):
-    file_name = getbundlefilename();
     file_directory = os.path.join(os.getcwd(), PackagesDirectory)
 
-    cmd = UninstallCommandTemplate.format(file_name)
-    ScriptUtil.run_command(hutil, ScriptUtil.parse_args(cmd), file_directory, 'Uninstall', ExtensionShortName, hutil.get_extension_version())
+    cmd = UninstallCommandTemplate.format(BundleFileName)
+    waagent.Log("Starting command %s --remove." %(BundleFileName))
 
+    # Retry up to three times, since uninstall can fail due to concurrent package operations
+    exit_code = run_command_with_retries(hutil, cmd, file_directory, 'Uninstall',
+                                         ExtensionShortName, retries = 10)
+    return exit_code
 
-def enable(hutil):
-    hutil.exit_if_enabled()
 
+def enable(hutil):
+    waagent.Log("Handler not enabled. Starting onboarding.")
     public_settings = hutil.get_public_settings()
     protected_settings = hutil.get_protected_settings()
     if public_settings is None:
-        raise ValueError("Public configuration cannot be None.")
+        raise ValueError("Public configuration must be provided")
     if protected_settings is None:
-        raise ValueError("Private configuration cannot be None.")
+        raise ValueError("Private configuration must be provided")
 
     workspaceId = public_settings.get("workspaceId")
     workspaceKey = protected_settings.get("workspaceKey")
+    proxy = protected_settings.get("proxy")
+    vmResourceId = protected_settings.get("vmResourceId")
+    stopOnMultipleConnections = public_settings.get("stopOnMultipleConnections")
     if workspaceId is None:
-        raise ValueError("Workspace ID cannot be None.")
+        raise ValueError("Workspace ID must be provided")
     if workspaceKey is None:
-        raise ValueError("Workspace key cannot be None.")
+        raise ValueError("Workspace key must be provided")
+
+    if stopOnMultipleConnections is not None and stopOnMultipleConnections is True:
+        output_file = tempfile.NamedTemporaryFile('w')
+        output_file.close()
 
-    cmd = OnboardCommandTemplate.format(workspaceId, workspaceKey)
-    exit_code = ScriptUtil.run_command(hutil, ScriptUtil.parse_args(cmd), OmsAdminWorkingDirectory, 'Onboard', ExtensionShortName, hutil.get_extension_version(), False)
-    # if onboard succeeds we continue, otherwise fail fast
+        list_exit_code = ScriptUtil.run_command(hutil,
+                                                ScriptUtil.parse_args(WorkspaceCheckCommand),
+                                                OmsAdminWorkingDirectory,
+                                                'Check If Already Onboarded', ExtensionShortName,
+                                                hutil.get_extension_version(), False,
+                                                interval = 30,
+                                                std_out_file_name = output_file.name)
```

Comment:
I now finally see why you need the tempfile module... Your better bet would be probably to use waagent.RunGetOutput(), rather than using ScriptUtil.run_command(). That way, you don't need to create a temporary file and redirect stdout to the temp file. waagent.RunGetOutput() returns the resulting stdout as a string. Lots of tribal knowledge... :)